opencascade: import from homebrew/science

bblacey opened the pull request:
Refs Homebrew/homebrew-science/pull/6495#issuecomment-351875336
Refs Homebrew/homebrew-science/issues/6331
Refs Homebrew/homebrew-core/issues/15619
History https://github.com/Homebrew/homebrew-science/commits/ba6c99f8d2154af8e1a50ed803c505dc7aefa5c4/opencascade.rb

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [X] Confirmed that tests pass locally (on High Sierra) `brew test opencascade`
-----

bblacey commented:
```
install events in the last 90 days for opencascade
================================================================================
1 | homebrew/science/opencascade                                 | 233 |  95.88%
2 | homebrew/science/opencascade --with-extras --with-test       |   7 |   2.88%
3 | homebrew/science/opencascade --with-extras                   |   2 |   0.82%
4 | homebrew/science/opencascade --without-opencl --with-extras  |   1 |   0.41%
```

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+
+  depends_on "cmake" => :build
+  depends_on "freetype"
+  depends_on "freeimage" => :recommended
```
Let's make `freeimage`, `tbb` and `gl2ps` mandatory.
Well remove the `test` option, as well as `without-opencl`.
For extras, we want to remove that option, I'd lean toward including docs (and thus build dependency on `doxygen`), but not source and the rest.

bblacey replied:
All great feedback however I don't think we should include the docs or extras because it increases the bottle size by nearly 1GB which wastes significant recurring resources on Travis and CircleCI (`opencascade` is a library/SDK).  I will do the following:

1. add the mandatory for `freeimage`, `tbb`, `gl2ps`
2. remove `without-opencl`
3. leave `test` and `extras` for now.  I will try to do some analysis but on bottle-size but it was a significant hit in the past and the Travis team has been reaching out to macOS users with continuous integration builds asking them to optimize.  For `FreeCAD`, I developed a [Homebrew ports cache](https://github.com/FreeCAD/FreeCAD-ports-cache) to reduce builds times well-beyond what bottled-installs achieve because the homebrew port installs alone were exceeding the standard Travis CI build time allocation.

bblacey replied:
removed `test` because the legacy hb/science formula had deprecated the `test` option and opencascade appears as a "new" formula in hb/core and the build-bot rejects new formula with deprecated options (makes sense)

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+  url "https://github.com/FreeCAD/homebrew-freecad/releases/download/0/opencascade-7.2.0.tgz"
+  sha256 "f9a81ca2267635773af9fff80625735dd4f65389af986fe93a94263081b422e9"
+
+  bottle do
```
Remove the bottle block, it will be regenerated automatically.

bblacey replied:
Per the instructions I read, it was my understanding that I was supposed to leave the bottle block so is this suggestion specific to porting a formula from science to core?  Will fix.

ilovezfs replied:
We've been inconsistent with this. It has no impact other than what message is used for the new bottles: "added" vs "updated"

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+  depends_on "tbb"       => :recommended
+  depends_on "gl2ps"     => :recommended
+  depends_on "doxygen"   if build.with? "extras"
+  depends_on :macos      => :snow_leopard
```
This is not necessary anymore.

bblacey replied:
Agreed.  Will fix.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+  depends_on "doxygen"   if build.with? "extras"
+  depends_on :macos      => :snow_leopard
+
+  conflicts_with "oce", :because => "OCE is a fork for patches/improvements/experiments over OpenCascade"
```
cross-tap conflicts don't work so this just needs to be removed

bblacey replied:
Thanks for the peer review - will fix.

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+      cmake_args << "-DUSE_#{feature.upcase}:BOOL=ON" if build.with? feature
+    end
+
+    sdk_path = Pathname.new `xcrun --show-sdk-path`.strip
```
`skd_path` can be replaced with `MacOS.sdk_path`

ilovezfs replied:
It's also not correct. This should be
```
sdk = MacOS::CLT.installed? ? "" : MacOS.sdk_path
```

bblacey replied:
Good catch

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+    cmake_args << "-D3RDPARTY_TCL_INCLUDE_DIR:PATH=#{sdk_path}/usr/include/"
+    cmake_args << "-D3RDPARTY_TK_INCLUDE_DIR:PATH=#{sdk_path}/usr/include/"
+    cmake_args << "-DCMAKE_FRAMEWORK_PATH:PATH=#{HOMEBREW_PREFIX}/Frameworks"
+    cmake_args << "-D3RDPARTY_FREETYPE_INCLUDE_DIR:PATH=#{HOMEBREW_PREFIX}/include/freetype2"
```
Use `#{Formula["freetype"].opt_prefix}/include/freetype2`

ilovezfs replied:
opt_include

bblacey replied:
Will do

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+    cmake_args << "-DCMAKE_FRAMEWORK_PATH:PATH=#{HOMEBREW_PREFIX}/Frameworks"
+    cmake_args << "-D3RDPARTY_FREETYPE_INCLUDE_DIR:PATH=#{HOMEBREW_PREFIX}/include/freetype2"
+    opencl_path = Pathname.new "#{sdk_path}/System/Library/Frameworks/OpenCL.framework/Versions/Current"
+    if build.with?("opencl") && opencl_path.exist?
```
This should always exist.

bblacey replied:
Agreed

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+
+  test do
+    ENV["CASROOT"] = opt_prefix
+    "1\n"==`#{bin}/DRAWEXE -c \"pload ALL\"`
```
What does that test do?
It will need to be reworked to use `assert_match` or another predicate.

bblacey replied:
This simply runs the DRAWEXE executable and loads the available modules to ensure opencascade built and installed correctly.  I will see if I can morph this test block into a more acceptable homebrew pattern.

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+
+  def install
+    cmake_args = std_cmake_args
+    cmake_args << "-DCMAKE_PREFIX_PATH:PATH=#{HOMEBREW_PREFIX}"
```
Wow, that's a lot of non-standard cmake variables. You sure we need them all?

bblacey replied:
Per instructions here - https://www.opencascade.com/doc/occt-7.2.0/overview/html/occt_dev_guides__building_cmake.html#build_cmake_build

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+  def install
+    cmake_args = std_cmake_args
+    cmake_args << "-DCMAKE_PREFIX_PATH:PATH=#{HOMEBREW_PREFIX}"
+    cmake_args << "-DCMAKE_INCLUDE_PATH:PATH=#{HOMEBREW_PREFIX}/lib"
```
Why does that `INCLUDE_PATH` contain `lib`? That seems wrong.

bblacey replied:
Will investigate

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+
+  def caveats; <<-EOF.undent
+    Some apps will require this enviroment variable:
+      CASROOT=#{opt_prefix}
```
Then we should use a wrapper script instead of a caveat.

bblacey replied:
I'm by no means a homebrew formula expert so can you provide an example or code snippet of how to accomplish this?

ilovezfs replied:
It still seems wrong to me that people should have to set this in their environment when we know it at installation time.

bblacey replied:
@MikeMcQuaid - I just returned from Holiday travel so I am a bit behind on this work-in-progress but am committed to finishing it.  Beyond the formula's built-in test, I also like to test the resulting opencascade install against FreeCAD as an acid test that covers more surface area and those iterations take a bit of time.

@ilovezfs That said, I plan to address all comments and can tackle wrappers but the upstream doesn't provide any wrapper scripts and ```find /usr/local/Homebrew/Library/Taps/homebrew -name '*.rb' -exec grep -H -i wrapper  {} \;``` didn't yield any examples.  Can you provide an example of what you would like to see?  Ideally an existing exemplar Formula that codifies your desires that I can use as an implementation reference.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+      end
+    end
+
+    # add symlinks to be able to compile against OpenCascade
```
if it's not installed by "make install" then we don't want to install it.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,91 @@
+      system "make", "install"
+
+      if build.with? "extras"
+        prefix.install "../src"
```
The fact that this is dumping things directly in `prefix` makes me think we should remove the extras option entirely.

bblacey commented:
To make it easier for your continued review, I will push individual commits that address the feedback here. Once you guys feel we are done, I will squash the commit history down to something more reasonable.

bblacey commented:
@ilovezfs, @fxcoudert, @sjackman - For reasons I fail to understand, the CI BuildBot considers personal GitHub notoriety a measure of one's technical competence/prowess and merit for the value of their contributions (I have made several to Homebrew in the past).  Sure, I could to game the system by creating a bunch of forks and/or marketing for stars but that is not a good use of anyone's time.  Unless this policy changes, I will not be able to finish this pull request nor be party to this because I think a CI check like this exudes an unwelcoming and arrogant attitude for an open source
project, especially when only admins/mods can merge pull requests.  Shockingly bad form guys.

```
Error: 1 problem in 1 formula
opencascade:
  * GitHub repository not notable enough (<20 forks, <20 watchers and <50 stars)
```

fxcoudert commented:
The notoriety is not about you, it's about the project: we want to host and maintain formulas for software that has a certain threshold of users.

In this particular case, the source should be downloaded from the primary/canonical URL, which seems to be `https://www.opencascade.com/sites/default/files/private/occt/OCC_7.2.0_release/opencascade-7.2.0.tgz`.

Also, the `url` should be the main entry page for users (`https://www.opencascade.com/content/core-technology` if I am not mistaken), not the dev portal.

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,73 @@
+    cmake_args << "-DUSE_FREEIMAGE:BOOL=ON"
+    cmake_args << "-DUSE_TBB:BOOL=ON"
+    cmake_args << "-D3RDPARTY_TBB_DIR:PATH=#{HOMEBREW_PREFIX}"
+    cmake_args << "-D3RDPARTY_TCL_DIR:PATH=/usr"
```
Wondering if this shouldn't be `#{sdk}/usr`, given the lines below

bblacey replied:
The legacy formula uses the macOS built-in tcl-tk so this path is correct.  

However, in the context of this comment, is there a Homebrew best-practice or standard for determining the precedence of using macOS built-in dependencies vs. Homebrew dependencies when developing formula?  For example, should formulae always depend upon Homebrew packages if a dependency is available in Homebrew and fallback to macOS packages or vice-versa?  What are your thoughts on using macOS tcl-tk or the homebrew tcl-tk?

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,73 @@
+    cmake_args << "-DUSE_GL2PS:BOOL=ON"
+    cmake_args << "-DUSE_FREEIMAGE:BOOL=ON"
+    cmake_args << "-DUSE_TBB:BOOL=ON"
+    cmake_args << "-D3RDPARTY_TBB_DIR:PATH=#{HOMEBREW_PREFIX}"
```
`#{Formula["tbb"].opt_prefix}`

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,73 @@
+    cmake_args = std_cmake_args
+    sdk = MacOS::CLT.installed? ? "" : MacOS.sdk_path
+    opencl_path = Pathname.new "#{sdk}/System/Library/Frameworks/OpenCL.framework/Versions/Current"
+    cmake_args << "-DCMAKE_PREFIX_PATH:PATH=#{HOMEBREW_PREFIX}"
```
The 3 `CMAKE_*_PATH` variables are not documented on https://www.opencascade.com/doc/occt-7.2.0/overview/html/occt_dev_guides__building_cmake.html#build_cmake_build
Probably they are not needed, and `std_cmake_args` already contains the necessary generic variables for setting prefix.

fxcoudert commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,73 @@
+    cmake_args << "-DCMAKE_PREFIX_PATH:PATH=#{HOMEBREW_PREFIX}"
+    cmake_args << "-DCMAKE_INCLUDE_PATH:PATH=#{HOMEBREW_PREFIX}/lib"
+    cmake_args << "-DCMAKE_FRAMEWORK_PATH:PATH=#{HOMEBREW_PREFIX}/Frameworks"
+    cmake_args << "-D3RDPARTY_DIR:PATH=#{HOMEBREW_PREFIX}"
```
We should find which formula this is looking for, and use `Formula["xxx"].opt_prefix` here also

bblacey commented:
@fxcoudert - thanks for the GitHub notoriety clarification - I'm glad that my understanding was incorrect.  That said, the formula can not download the opencascade source tarball from the opencascade site without an account.  That is why other package managers like conda forge and archlinix stage the download elsewhere - e.g. https://github.com/conda-forge/occt-feedstock/blob/master/recipe/meta.yaml and why I uploaded it to FreeCAD (as we have done in the past).  Does Homebrew have a similar mechanism available to host source downloads?  If not, do you have any other ideas? 

The opencascade formula is not new, I am simply lending a hand to port it from science to core so much of what is here, existed previously, perhaps for linuxbrew compatibility.  It makes sense to revamp the formula to bring it up to macOS-only core standards and with your feedback/critique, we'll get there.  I'll work to trim the legacy `cmake` options to the minimum beyond `std_cmake_args`.

ilovezfs commented:
>That said, the formula can not download the opencascade source tarball from the opencascade site without an account.

That smells like a licensing issue to me.

ilovezfs commented:
Looks like it's LGPL so that's OK. But
```
$ sha256 opencascade-7.2.0.tgz 
SHA256(opencascade-7.2.0.tgz)= 245e344b2cf8a6e0a7d9f45d5d807f8ef48607bcd7a6d763ba2d7b861fa55249
```
for the one from their site.

bblacey commented:
I used `curl -L -o opencascade-7.2.0.tgz "http://git.dev.opencascade.org/gitweb/?p=occt.git;a=snapshot;h=refs/tags/V7_2_0;sf=tgz"`

ilovezfs commented:
That url gives me a different checksum every time.

Unfortunately, until there's an official upstream url for the release tarball that doesn't require login credentials, we're not going to be able to package this software in homebrew/core. You'll need to host this in your own tap for the time being.

Thanks for the PR nonetheless @bblacey.

bblacey commented:
Understood.  I will host it in our FreeCAD-homebrew tap for now. I will also contact the opencascade folks to see if they can offer an option for Homebrew to offer it more broadly. 

Thanks for all your help - always a learning experience.

ilovezfs commented:
@bblacey thanks for understanding. I hope they can work something out for us.

bblacey commented:
@ilovezfs - I just sent a note to the project manager for the opencascade kernel.  Hopefully he will be able to help directly or point me to someone who can.  Fingers-crossed that we can find an acceptable solution so the formula can be more generally available to the masses.

bblacey commented:
@ilovezfs, I realize this PR is closed but I wanted to provide an update....  From the project manager for the opencascade kernel - looks like they are interested in helping find a solution.

>Hello Bruce,
>Thank you for your message!
>Indeed we are aware of Homebrew and would be happy to see OCCT integrated.
>Unfortunately, by the time we still have no standard public interface for anonymous access to OCCT Git.
>I will try to reiterate on this and expect to have update of the status by Friday, then will let you know.
>
>Best Regards,
>Andrey

ilovezfs commented:
@bblacey thanks for the update. It would be good to have access to the actual release tarball as well. Using a Git repository directly for something other than a --HEAD spec is a last resort when a build system fails without direct access to Git history.

bblacey commented:
Agreed.  Andrey actually suggested, in additional details, that we use git snapshots in the interim but I told them that they don't currently work due to varyings shas and that we really need a url for the source archive and companion sha hosted on an official opencascade site.  Hopefully we will know more by the end of the week.

bblacey commented:
@ilovezfs - Andrey the Opencascade kernel project lead is very supportive of Homebrew integration and has been extremely responsive and helpful.  It turns out there was a bug in the opencascade gitweb deployment that caused the sha mismatch that has been corrected.  Andrey also confirmed that downloading the source tarball using their gitweb server complies with their licensing terms and is the preferred means for downloading the source.  Here are the relevant excerpts from our email exchanges:

>Hello Bruce, it was actually a bug in GitWeb that resulted in different snapshot archives in tgz format. Now it should be fixed, so you can re-try with tgz.
>Please let me know if this solved the issue.
>Best Regards,
>Andrey
---
>Hi Andrey, Thanks for all your help and for fixing the gitweb issue so quickly.  I have confirmed that your fix yields the same sha. I want to confirm that this is an approved means for providing Homebrew integration.  In other words, we will be in compliance with your licensing terms if we use the gitweb interface to build from sources for the Homebrew formula/distribution.
Thanks,
Bruce
---
>Hello Bruce,
>
>Using GitWeb to download source snapshots is fully compliant with our rules, and has always been a recommended way to access OCCT for people who needs more than official release but do not have or do not need full access to Git. (The standard way of downloading official version still requires authentication -- we need this to have >at least partial statistics on OCCT user base.)
>The content of the archives is actually different: in addition to contents of the Git repo, source package in official distribution (https://www.opencascade.com/content/latest-release) contains also many generated files (MS Visual Studio, Code Blocks, and XCode projects, scripts, and documentation). (Note different sizes: 45 MB of Git snapshot vs. 193 MB of the official package.)
In addition, tgz files produced by GitWeb contain additional sub-folder inside and are apparently packaged by different version of gzip. Apart of this, there should be no difference, thus the libraries built from both packages should be identical.
>As for the long-term accessibility, having URL pointing to GitWeb looks even more reliable than official download link.
The latter can be changed if web site is reorganized, while GitWeb is hosted on separate server and is not going to change (at least, it works since 2011 in the same place and we have no plans to reformat it).
>I will be looking forward to news from you.
>
>Best Regards,
>Andrey

I'de like to continue refining the legacy opencascade formula into Core compliance based upon your and @fxcoudert's review comments if you agree that the gitweb download that Andrey/OpenCascade provided is acceptable.  It will take a bit of time to iterate and test offline so I want to confirm that you are amenable to this before investing additional effort and submitting a new pull request.

MikeMcQuaid commented:
@bblacey Nice work so far and thanks for your patience. Could you also tweak the `homebrew/science/opencascade` to be just `opencascade` in the `gmsh` formula? Thanks!

ilovezfs commented:
@bblacey as long as the checksum remains stable and the tarball has all the bits you need for the formula, that should be fine :)

bblacey commented:
@MikeMcQuaid - sure, will do.
@ilovezfs - thanks!  I'll push a commit with the updated sha before addressing some of the other areas so we can see if it will build across OSes.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,67 @@
+class Opencascade < Formula
+  desc "3D modeling and numerical simulation software for CAD/CAM/CAE"
+  homepage "https://www.opencascade.com/content/core-technology"
+  url "http://git.dev.opencascade.org/gitweb/?p=occt.git;a=snapshot;h=refs/tags/V7_2_0;sf=tgz"
```
You'll need to set the version manually.

```
Josephs-MacBook-Pro:~ joe$ brew irb
Ve==> Interactive Homebrew Shell
Example commands available with: brew irb --examples
irb(main):001:0> Version.detect("http://git.dev.opencascade.org/gp=occt.git;a=snapshot;h=refs/tags/V7_2_0;sf=tgz", {})
=> #<Version::FromURL:0x00007fb4b148cf20 @version="2">
irb(main):002:0>
```
So
```
version "7.2.0"
```

bblacey replied:
Will do.  Thanks for the irb tip/tutorial too.

bblacey replied:
Question - have you guys considered adding a version validity check (e.g. semantic version pattern) to `brew audit --strict`?

ilovezfs replied:
No but that is a good idea if you want to open a PR to Homebrew/brew

bblacey replied:
Perhaps something like this would work?
```
[blacey@bbl /usr/local/Homebrew/Library/Homebrew/dev-cmd (master *)]$ git diff
diff --git a/Library/Homebrew/dev-cmd/audit.rb b/Library/Homebrew/dev-cmd/audit.rb
index 89ab79996..671133cb2 100644
--- a/Library/Homebrew/dev-cmd/audit.rb
+++ b/Library/Homebrew/dev-cmd/audit.rb
@@ -865,8 +865,8 @@ class ResourceAuditor
       problem "version #{version} should not have a leading 'v'"
     end
 
-    return unless version.to_s =~ /_\d+$/
-    problem "version #{version} should not end with an underline and a number"
+    return if version.to_s =~ /^HEAD$|^DEVEL$|^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(\.(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*)?(\+[0-9a-zA-Z-]+(\.[0-9a-zA-Z-]+)*)?$/
+    problem "version \"#{version}\" must conform to semantic versioning requirements"
   end
 
   def audit_download_strategy
```

Example 1 (removed the Version declaration):
```
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master *)]$ brew audit --strict opencascade
opencascade:
  * Stable: version "2" must conform to semantic versioning requirements
  * stable version should not decrease (from 7.2.0 to 2)
Error: 2 problems in 1 formula
```

Example 2 (add Version "7.2.0")
```
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master)]$ brew audit --strict opencascade
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master)]$
```

That said, this could be a disruptive change if there are many existing formula that don't currently comply with semantic versioning requirements.

MikeMcQuaid commented:
@bblacey Gentle ping on the comments here when you get a chance ❤️

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+        prefix.install "../adm"
+        share.install_symlink prefix/"adm"
+      else
+        # Some apss expect resoures in legacy ${CASROOT}/src directory
```
spelling

bblacey replied:
Legacy formula defect, will fix

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+  end
+
+  test do
+    ENV["CASROOT"] = opt_prefix
```
we should at least be able to use wrapper scripts for the binaries that come with opencascade

bblacey replied:
I need a bit of additional guidance to make this happen - see https://github.com/Homebrew/homebrew-core/pull/21773#discussion_r159080756

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+  sha256 "adb6fc44efebdb09e30a487fa412afae9599992dc771b0039b4ebf306891a558"
+  head "http://git.dev.opencascade.org/gitweb/?p=occt.git;a=snapshot;h=HEAD;sf=tgz"
+
+  option "with-extras", "Install documentation (~17 MB), source files (~113 MB), samples and templates"
```
don't think we need this option

bblacey replied:
Opencascade is in essence an SDK or, Kit in Apple parlance.  This means that there are two potential use cases or scenarios in the Homebrew ecosystem to consider. 

*CAD User* - Homebrew users installing a program/package that uses the opencascade SDK libraries
*CAD Developer* - Homebrew users who are developing an application that uses opencascade 

The formula defaults satisfy the *CAD User* however *CAD Developers* require access to the API documentation and source code hence the legacy `--with-extras` developer-convenience option.  

It is clear from the `brew analytics` above that `--with-extras` is rarely used but if we drop `--with-extras`, we will force *CAD Developers* to manually install/configure opencascade (they can use the formula as a reference but it certainly isn't convenient) outside the Homebrew ecosystem.

If everyone agrees that we don't need to support the *CAD Developer* use case, then, at a minimum, we should install the documentation as @fxcoudert suggested at the small of cost of increased bottle size.

I'm not advocating one or the other, I'm simply sharing context for us to make an informed choice/decision.  Please advise and I will make it happen.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+  def install
+    cmake_args = std_cmake_args
+    sdk = MacOS::CLT.installed? ? "" : MacOS.sdk_path
+    cmake_args << "-DCMAKE_FRAMEWORK_PATH:PATH=#{HOMEBREW_PREFIX}/Frameworks"
```
are you sure this is needed?

bblacey replied:
On my list to confirm.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+    cmake_args << "-DUSE_GL2PS:BOOL=ON"
+    cmake_args << "-DUSE_FREEIMAGE:BOOL=ON"
+    cmake_args << "-DUSE_TBB:BOOL=ON"
+    cmake_args << "-D3RDPARTY_TBB_DIR:PATH=#{HOMEBREW_PREFIX}"
```
should use Formula["tbb"].opt_prefix

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+    cmake_args << "-DUSE_TBB:BOOL=ON"
+    cmake_args << "-D3RDPARTY_TBB_DIR:PATH=#{HOMEBREW_PREFIX}"
+    cmake_args << "-D3RDPARTY_TCL_DIR:PATH=/usr"
+    cmake_args << "-D3RDPARTY_TCL_INCLUDE_DIR:PATH=#{sdk}/usr/include/"
```
the trailing slash shouldn't be needed

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+    cmake_args << "-D3RDPARTY_TBB_DIR:PATH=#{HOMEBREW_PREFIX}"
+    cmake_args << "-D3RDPARTY_TCL_DIR:PATH=/usr"
+    cmake_args << "-D3RDPARTY_TCL_INCLUDE_DIR:PATH=#{sdk}/usr/include/"
+    cmake_args << "-D3RDPARTY_TK_INCLUDE_DIR:PATH=#{sdk}/usr/include/"
```
ditto

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+  depends_on "doxygen" if build.with? "extras"
+
+  def install
+    cmake_args = std_cmake_args
```
variable should be named `args`

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,68 @@
+    lib.install_symlink Dir["#{loc}/lib/*"]
+  end
+
+  def caveats; <<-EOF.undent
```
if we do keep the caveats, this needs to use the new style squiggly heredocs

bblacey replied:
I think we should keep it.  I will change it to squiggly heredocs...

bblacey commented:
@ilovezfs, @fxcoudert and @MikeMcQuaid, I believe I have addressed all of the review comments sans wrappers because I need an exemplar to follow - see https://github.com/Homebrew/homebrew-core/pull/21773#discussion_r159080756.  If you want to forego wrappers at this time, then feel free to squash the commit history on merge or ping me to do it.  Apologies for not using a branch for this PR - my bad.

Thanks for all the great feedback!

ilovezfs commented:
Wrapper examples:
https://github.com/Homebrew/homebrew-core/blob/5ad6d0d69206945bcbf5f6db0bf3abb3246a967a/Formula/st.rb#L21-L22
https://github.com/Homebrew/homebrew-core/blob/master/Formula/sbt.rb#L20-L27
https://github.com/Homebrew/homebrew-core/blob/master/Formula/git-ftp.rb#L45-L46
https://github.com/Homebrew/homebrew-core/blob/master/Formula/metabase.rb#L27-L31

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+  depends_on "cmake" => :build
+  depends_on "freetype"
+  depends_on "freeimage"
+  depends_on "tbb"
```
non-build deps need to be alphabetized amongst themselves

bblacey replied:
Will fix.  Seems like another candidate for `brew audit --strict` ;)

ilovezfs replied:
not really because there are cases where that's not actually possible

bblacey replied:
too bad because I'm sure you are getting tired of telling *newbs* to alphabetize their depends ;)

ilovezfs replied:
I find they usually learn the alphabet relatively quickly.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+
+  def install
+    args = std_cmake_args
+    args << "-DUSE_GL2PS:BOOL=ON"
```
nothing is appended optionally so these should all be inlined below as in https://github.com/Homebrew/homebrew-core/blob/master/Formula/corsixth.rb#L48-L51

bblacey replied:
done

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+      system "make", "install"
+    end
+
+    # Some apps expect resources in legacy ${CASROOT}/src directory
```
if make install doesn't create such a symlink, we shouldn't either

bblacey replied:
This was in the original opencascade formula and the binaries won't work without it or at least I don't know how to make it work without it...

ilovezfs replied:
Then this needs to be reported upstream and we need a link to the issue report here because `make install` should create this symlink if it's needed for a functional installation.

bblacey replied:
Sounds good.  I will send a note to Andrey at opencascade to find out what I need to do to make that happen.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+    end
+  end
+
+  def caveats; <<~EOS
```
this seems like the purview of upstream documentation. If we can't handle this automatically, let's just skip the caveat

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+
+  test do
+    ENV["CASROOT"] = opt_prefix
+    assert_match "1", shell_output("#{bin}/DRAWEXE -c \"pload ALL\"")
```
can we use assert_equal?

bblacey replied:
@fxcoudert recommended `assert_match` in an earlier because the legacy formula didn't even use `assert`.  What is the benefit of `assert_equal` over `assert_match`?

ilovezfs replied:
`assert_match` is for when we can only easily match a regex or a substring. `assert_equal` is more precise and sometimes requires just a `.chomp` on the end of the `shell_output`

bblacey replied:
done

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+    args << "-DUSE_TBB:BOOL=ON"
+    args << "-DINSTALL_DOC_Overview:BOOL=ON"
+
+    mkdir "build" do
```
does an in-tree build work?

bblacey replied:
Opencascade recommends and out-of-souce build.  An earlier version of the legacy formula did an in-tree build that resulted in in all the intermediate binaries being included in the bottle.

ilovezfs replied:
ok but does that still happen?

bblacey replied:
Don't know - that is from recollection circa Opencasde 6.x.  I can regress to see but what is the advantage of an in-tree build?

ilovezfs replied:
just two fewer lines. Fewer lines = betterer

bblacey replied:
LoL - I will see if I can do betterer ;)

bblacey replied:
Outside source-tree build:
```
==> Not running post_install as we're building a bottle
You can run it manually using `brew postinstall opencascade`
==> Summary
🍺  /usr/local/Cellar/opencascade/7.2.0: 9,235 files, 164.6MB, built in 14 minutes 3 seconds
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master)]$ brew bottle opencascade
==> Determining opencascade bottle rebuild...
==> Bottling opencascade-7.2.0.high_sierra.bottle.1.tar.gz...
==> Detecting if opencascade-7.2.0.high_sierra.bottle.1.tar.gz is relocatable... 
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master)]$ ls -lh opencascade-7.2.0.high_sierra.bottle.1.tar.gz                    
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master)]$ ls -lh opencascade-7.2.0.high_sierra.bottle.1.tar.gz 
-rw-r--r--  1 blacey  admin    50M Aug 31 05:47 opencascade-7.2.0.high_sierra.bottle.1.tar.gz
```

In source-tree build:
```
==> Not running post_install as we're building a bottle
You can run it manually using `brew postinstall opencascade`
==> Summary
🍺  /usr/local/Cellar/opencascade/7.2.0: 9,235 files, 164.6MB, built in 14 minutes 35 seconds
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master *)]$ brew bottle opencascade
==> Determining opencascade bottle rebuild...
==> Bottling opencascade-7.2.0.high_sierra.bottle.1.tar.gz...
==> Detecting if opencascade-7.2.0.high_sierra.bottle.1.tar.gz is relocatable... 
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master *)]$ ls -lh opencascade-7.2.0.high_sierra.bottle.1.tar.gz
-rw-r--r--  1 blacey  admin    50M Aug 31 05:47 opencascade-7.2.0.high_sierra.bottle.1.tar.gz
```

Bottom-line, looks like an *in source-tree* build yields the same results so I will change the formula to be *betterer* ;)

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+  depends_on "tbb"
+  depends_on "gl2ps"
+  depends_on "tcl-tk"
+  depends_on "doxygen"
```
surely this is build time only

bblacey replied:
Yep.  Only needed to build and install the docs.

ilovezfs replied:
then `=> :build`

bblacey replied:
done

ilovezfs commented:
Can you please
```
brew unlink doxygen freeimage freetype gl2ps libpng tbb tcl-tk
```
and make sure the build still works (this is how to check if non-/usr/local prefixes will work)

If not, we should pass the necessary cmake paramaters to make it work when everything is unlinked.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+  depends_on "freeimage"
+  depends_on "tbb"
+  depends_on "gl2ps"
+  depends_on "tcl-tk"
```
why can't we use macOS's?

bblacey replied:
We can.  I asked for guidance on that here https://github.com/Homebrew/homebrew-core/pull/21773#discussion_r159081797 and thought there might be a preference on Homebrew over macOS.  Please advise.

ilovezfs replied:
With the exception of `openssl` and occasionally `python`, we always want to use what macOS provides if possible.

bblacey replied:
done

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+  def install
+    args = std_cmake_args
+    args << "-DUSE_GL2PS:BOOL=ON"
+    args << "-DUSE_FREEIMAGE:BOOL=ON"
```
we don't need to explicitly specify `:BOOL`, which is redundant

bblacey replied:
doe

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  test do
+    ENV["CASROOT"] = opt_prefix
```
this shouldn't be needed once you have the wrappers

bblacey replied:
yep

bblacey commented:
@ilovezfs - per your request:

>Can you please
>`brew unlink doxygen freeimage freetype gl2ps libpng tbb tcl-tk`

Sure enough, the opencascade cmake config is not able to find the packages when unlinked so I have added explicit paths for the dependencies.opt_prefix.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -7,39 +7,43 @@ class Opencascade < Formula
+                            "-DUSE_FREEIMAGE=ON",
+                            "-DUSE_TBB=ON",
+                            "-DINSTALL_DOC_Overview=ON",
+                            "-D3RDPARTY_FREETYPE_DIR=#Formula['freetype'].opt_prefix",
```
```
"-D3RDPARTY_FREETYPE_DIR=#{Formula["freetype"].opt_prefix}",
```

bblacey replied:
D'oh - was getting tired.  Fixed.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -7,39 +7,43 @@ class Opencascade < Formula
       system "make", "install"
     end
 
+    # Wrapper
```
I think you want
```
bin.env_script_all_files(libexec/"bin", :CASROOT => prefix)
```

bblacey replied:
How do I setup the `brew irb` context so I can I test this using `brew irb`?  `brew irb --examples` fails on my machine.

ilovezfs replied:
```
bin = Formula["opencascade"].bin
```

ilovezfs replied:
```
libexec = Formula["opencascade"].libexec
```

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -7,39 +7,43 @@ class Opencascade < Formula
   test do
-    ENV["CASROOT"] = opt_prefix
-    assert_match "1", shell_output("#{bin}/DRAWEXE -c \"pload ALL\"")
+    assert_equal "1", shell_output("#{bin}/opencascade DRAWEXE -c \"pload ALL\"").chomp
```
```
    output = shell_output("#{bin}/opencascade DRAWEXE -c \"pload ALL\"")
    assert_equal "1", output.chomp
```

bblacey replied:
I thought fewer lines == betterer? ;)  Fixed.

ilovezfs replied:
not when they go passed the eightieth column :)

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,49 @@
+      CASROOT=#{opt_prefix} exec #{opt_prefix}/bin/"$@"
+    EOS
+
+    # Some apps expect resources in legacy ${CASROOT}/src directory
```
```
prefix.install_symlink pkgshare/"resources" => "src"
```
should work

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,49 @@
+
+    # Some apps expect resources in legacy ${CASROOT}/src directory
+    cd prefix do
+      ln_s "share/opencascade/resources", "src"
```
```
prefix.install_symlink pkgshare/"resources" => "src"
```
should work

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,49 @@
+                            "-D3RDPARTY_FREEIMAGE_DIR=#Formula['freeimage'].opt_prefix",
+                            "-D3RDPARTY_FREEGL2PS_DIR=#Formula['gl2ps'].opt_prefix",
+                            "-D3RDPARTY_TBB_DIR=#Formula['tbb'].opt_prefix",
+                            "-D3RDPARTY_TCL_DIR:PATH=/usr",
```
`:PATH` is redundant

bblacey replied:
Fair-enough.  I have vi setup to highlight column 80 and will dutifully oblige it's warnings when working on Homebrew formula. 

That said, while `brew audit --ilovezfs` works great, is this by chance another `brew audit --strict` opportunity?

ilovezfs replied:
If only it were always possible to stay <= 80 … The actual enforced maximum by `brew style` right now is 244 (:see_no_evil:) so you're stuck with `--ilovezfs`.

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,49 @@
+  url "http://git.dev.opencascade.org/gitweb/?p=occt.git;a=snapshot;h=refs/tags/V7_2_0;sf=tgz"
+  version "7.2.0"
+  sha256 "adb6fc44efebdb09e30a487fa412afae9599992dc771b0039b4ebf306891a558"
+  head "http://git.dev.opencascade.org/gitweb/?p=occt.git;a=snapshot;h=HEAD;sf=tgz"
```
Let's leave this out

bblacey commented:
@ilovezfs or @MikeMcQuaid - could one of you gents resolve the `gmsh` merge conflict or advise me on what I can should do to resolve so the CI build will continue?  Thanks.

ilovezfs commented:
What's the output of `git remote`

bblacey commented:
@ilovezfs 
```
[blacey@bbl /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula (master)]$ git remote -v -v
mbp13	ssh://blacey@bblmbp.local/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core (fetch)
mbp13	ssh://blacey@bblmbp.local/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core (push)
origin	https://github.com/bblacey/homebrew-core (fetch)
origin	https://github.com/bblacey/homebrew-core (push)
upstream	https://github.com/Homebrew/homebrew-core (fetch)
upstream	https://github.com/Homebrew/homebrew-core (push)
```

ilovezfs commented:
```
mv Formula/opencascade.rb ~/
git fetch upstream
git reset --hard upstream/master
mv ~/opencascade.rb Formula/
git add Formula/opencascade.rb
git commit -m "opencascade: import from homebrew/science"
git push -f
```

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+             *std_cmake_args
+    system "make", "install"
+
+    # Wrappers
```
comment not needed

bblacey replied:
done

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+  end
+
+  test do
+    output=shell_output("#{bin}/DRAWEXE -c \"pload ALL\"")
```
output=shell -> output = shell

bblacey replied:
done

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+    bin.env_script_all_files(libexec/"bin", :CASROOT => prefix)
+
+    # Some apps expect resources in legacy ${CASROOT}/src directory
+    cd prefix do
```
cd prefix do
  prefix.install_symlink pkgshare/"resources" => "src"
end

->

prefix.install_symlink pkgshare/"resources" => "src"

bblacey replied:
done

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+  def install
+    sdk = MacOS::CLT.installed? ? "" : MacOS.sdk_path
+    system "cmake", ".",
+             "-DUSE_GL2PS=ON",
```
this should be up on the same line as `system "cmake", ".",`

bblacey replied:
done

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+    sdk = MacOS::CLT.installed? ? "" : MacOS.sdk_path
+    system "cmake", ".",
+             "-DUSE_GL2PS=ON",
+             "-DUSE_FREEIMAGE=ON",
```
with the rest aligned beneath

bblacey replied:
If I start on the same line and use the same indent, some lines exceed 80 characters?  Which is it?

ilovezfs replied:
The former. There has to be room to doodle a snowman in the margin on the left.

bblacey replied:
Done - doodle away ;)

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,45 @@
+             "-DUSE_TBB=ON",
+             "-DINSTALL_DOC_Overview=ON",
+             "-D3RDPARTY_FREETYPE_DIR=#{Formula["freetype"].opt_prefix}",
+             "-D3RDPARTY_FREEIMAGE_DIR=#{Formula["freeimage"].opt_prefix}",
```
can we get these in the same order as the dependency listing

bblacey replied:
done

ilovezfs commented on Formula/opencascade.rb:
```diff
@@ -0,0 +1,41 @@
+                         "-DINSTALL_DOC_Overview=ON",
+                         "-D3RDPARTY_FREETYPE_DIR=#{Formula["freetype"].opt_prefix}",
+                         "-D3RDPARTY_FREEIMAGE_DIR=#{Formula["freeimage"].opt_prefix}",
+                         "-D3RDPARTY_FREEGL2PS_DIR=#{Formula["gl2ps"].opt_prefix}",
```
This should be `"-D3RDPARTY_GL2PS_DIR=#{Formula["gl2ps"].opt_prefix}",`

bblacey replied:
Stepped out for a burger and a beer - will address as some as I get home.  Sorry I botched that!

ilovezfs commented:
hmm locally I'm getting
```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
Appkit_LIB
    linked by target "TKService" in directory /tmp/opencascade-20171230-55191-ly5l23/occt-V7_2_0/src/TKService
    linked by target "TKOpenGl" in directory /tmp/opencascade-20171230-55191-ly5l23/occt-V7_2_0/src/TKOpenGl
    linked by target "TKDraw" in directory /tmp/opencascade-20171230-55191-ly5l23/occt-V7_2_0/src/TKDraw
    linked by target "TKViewerTest" in directory /tmp/opencascade-20171230-55191-ly5l23/occt-V7_2_0/src/TKViewerTest
```

ilovezfs commented:
My guess is that it's a case-sensitivity bug because my local file system is case sensitive.

ilovezfs commented:
Yup
```
    inreplace "adm/cmake/occt_csf.cmake", "Appkit", "AppKit"
```
fixes it.

bblacey commented:
@ilovezfs - do you want me to include the `inreplace` in the Formula to address the case-sensivity problem and then file an issue upstream?

bblacey commented:
@ilovezfs the latest commit should work correctly.  Changing `inreplace` to only change the `find_library (Appkit_LIB NAMES Appkit)` definition (i.e. not all instances of the Appkit string), worked locally for me.  Please give it a shot if/when you have some time.  Thanks again for all your help.

ilovezfs commented:
Yup that does the trick. Nice work!

>and then file an issue upstream

Yes please, and while you're at it, please request https for http://git.dev.opencascade.org. It's rather unfortunate the homepage has https and the url for the download doesn't.

One remaining issue here is that opencascade itself doesn't work while unlinked.

```
iMac-TMP:homebrew-core joe$ brew test opencascade -f
Testing opencascade
==> /usr/local/Cellar/opencascade/7.2.0/bin/DRAWEXE -c "pload ALL"
Error: opencascade: failed
<"1"> expected but was
<"An exception was caught 0x7ff88d03f010 : Draw_Failure: Could not open: TKTopTest; reason: dlopen(libTKTopTest.dylib, 1): image not found\n** Exception ** 0x7ff88d03f010 : Draw_Failure: Could not open: TKTopTest; reason: dlopen(libTKTopTest.dylib, 1): image not found">.

diff:
- 1
+ An exception was caught 0x7ff88d03f010 : Draw_Failure: Could not open: TKTopTest; reason: dlopen(libTKTopTest.dylib, 1): image not found
+ ** Exception ** 0x7ff88d03f010 : Draw_Failure: Could not open: TKTopTest; reason: dlopen(libTKTopTest.dylib, 1): image not found

folded diff:
- 1
+ An exception was caught 0x7ff88d03f010 : Draw_Failure: Could not open: TKTopTe
+ st; reason: dlopen(libTKTopTest.dylib, 1): image not found
+ ** Exception ** 0x7ff88d03f010 : Draw_Failure: Could not open: TKTopTest; reas
+ on: dlopen(libTKTopTest.dylib, 1): image not found
iMac-TMP:homebrew-core joe$ 
```

This is caused by the use of `dlopen` in `src/OSD/OSD_SharedLibrary.cxx` without a full path to the library, so the formula is dependent on the prefix being /usr/local or the user setting `DYLD_LIBRARY_PATH`.

We could do something like this:
```
diff --git a/Formula/opencascade.rb b/Formula/opencascade.rb
index 064c1f4901..93d87f5ada 100644
--- a/Formula/opencascade.rb
+++ b/Formula/opencascade.rb
@@ -12,7 +12,11 @@ class Opencascade < Formula
   depends_on "gl2ps"
   depends_on "tbb"
 
+  patch :DATA
+
   def install
+    inreplace "src/OSD/OSD_SharedLibrary.cxx", "@OPENCASCADE_PREFIX@", prefix
+
     inreplace "adm/cmake/occt_csf.cmake", "NAMES Appkit", "NAMES AppKit"
     sdk = MacOS::CLT.installed? ? "" : MacOS.sdk_path
     system "cmake", ".", "-DUSE_FREEIMAGE=ON",
@@ -40,3 +44,28 @@ class Opencascade < Formula
     assert_equal "1", output.chomp
   end
 end
+
+__END__
+diff --git a/src/OSD/OSD_SharedLibrary.cxx b/src/OSD/OSD_SharedLibrary.cxx
+index b635baa..4ff3f9a 100644
+--- a/src/OSD/OSD_SharedLibrary.cxx
++++ b/src/OSD/OSD_SharedLibrary.cxx
+@@ -108,11 +108,16 @@ void  OSD_SharedLibrary::SetName(const Standard_CString aName)  {
+ //
+ // ----------------------------------------------------------------
+ Standard_Boolean  OSD_SharedLibrary::DlOpen(const OSD_LoadMode aMode ) {
++
++char myFullPath[256];
++strlcpy(myFullPath, "@OPENCASCADE_PREFIX@/lib/", 256);
++strlcat(myFullPath, myName, 256);
++
+ if (aMode == OSD_RTLD_LAZY){
+-  myHandle = dlopen (myName,RTLD_LAZY);
++  myHandle = dlopen (myFullPath,RTLD_LAZY);
+ }
+ else if (aMode == OSD_RTLD_NOW){
+-  myHandle = dlopen (myName,RTLD_NOW);
++  myHandle = dlopen (myFullPath,RTLD_NOW);
+ }
+ 
+ if (!BAD(myHandle)){
```
But it would be far preferable to be able to pass the path to `cmake` or even just have it automatically look for the library in the `lib` directory if `dlopen` fails to find the library just by name.

We can look the other way on this for now but it would be good to get it reported upstream as this software really shouldn't be relying on the fact that /usr/local/lib happens to be in the default search path.

ilovezfs commented:
Thanks for the great work @bblacey! Shipped :tada:

bblacey commented:
Sweet!  I have reported the following 4 items to upstream:
1) Eliminate need for `prefix.install_symlink pkgshare/"resources" => "src"` required by legacy apps
2) Opencascade dynamic library prefix/path assumption/dependency
3) Building on Case-sensitive file systems (Appkit -> AppKit defect)
4) git.dev.opencascade.org over https (it seems they support ssl but the Let's Encrypt cert uses the wrong domain/host name so it fails ssl validation)

Also, is there anything that I need to do to remove `opencascade` from homebrew/science?

Finally, I still need to open a `brew` PR to add the semantic version check see in `brew audit --strict` - https://github.com/Homebrew/homebrew-core/pull/21773#discussion_r158586175.  I'll do it but I'm not sure it will fly with the broader Homebrew team because the number of Formulae do not currently use semantic versions will present a barrier but maybe I'll be surprised ;)

@ilovezfs, I appreciate your high standards and attention to detail that clearly require a lot of patient persistence.  Homebrew rocks and it is people like you who make it all happen. Thanks!

ilovezfs commented:
@bblacey Thank you for reporting those issues and the kind words!

>Also, is there anything that I need to do to remove opencascade from homebrew/science?

already done in https://github.com/Homebrew/homebrew-science/commit/f3ba1086e7050f99932cb6c1105685a4e3fbfd62

>Finally, I still need to open a brew PR to add the semantic version check see in brew audit --strict - #21773 (comment). I'll do it but I'm not sure it will fly with the broader Homebrew team because the number of Formulae do not currently use semantic versions will present a barrier but maybe I'll be surprised ;)

You might be able to do something fancy like check if the previous version matched the semantic versioning pattern, then enforce it for the new version, but if the previous version didn't match the semantic versioning pattern, then don't enforce it for the new version.

ilovezfs commented:
@bblacey https://github.com/Homebrew/brew/pull/2601 does something very similar to that.
